From a8d1f5679138f67c74f42165894f20bcd17f018b Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Wed, 2 Jul 2025 11:39:15 +0200
Subject: [PATCH 906/954] feat(stm32mp2): set DDR in self refresh in system off
 sequence

For power off sequence, force DDR in self refresh to deactivate PORT
cleanly and to avoid to have transaction on the bus which block Standby
entry.

The DDRSHR bit is also clear only when the DDR is no more used.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 3b88ad2f8b19759e5d98efe1acaba02d5b2158e6 feat(stm32mp2): set DDR in self refresh in system off sequence]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: Ie8650088a250afdac8944d23e9be02f607293ad0
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/473800
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 plat/st/stm32mp2/plat_ddr.c    | 5 +++++
 plat/st/stm32mp2/stm32mp2_pm.c | 4 +++-
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/plat/st/stm32mp2/plat_ddr.c b/plat/st/stm32mp2/plat_ddr.c
index 818193402c..080811c23e 100644
--- a/plat/st/stm32mp2/plat_ddr.c
+++ b/plat/st/stm32mp2/plat_ddr.c
@@ -311,6 +311,11 @@ int stm32mp_board_ddr_power_off(void)
 	 */
 	fixed_regulator_register();
 	initialize_pmic();
+
+	/* Force DDR in self refresh */
+	ddr_set_sr_mode(DDR_SSR_MODE);
+	ddr_sr_entry(true);
+
 	/* DDR and device tree is no more accessible when clk is deactivated */
 	ddr_sub_system_clk_off();
 	ddr_power_off(fdt, node);
diff --git a/plat/st/stm32mp2/stm32mp2_pm.c b/plat/st/stm32mp2/stm32mp2_pm.c
index 3ac1a0982a..2ae7a2cc50 100644
--- a/plat/st/stm32mp2/stm32mp2_pm.c
+++ b/plat/st/stm32mp2/stm32mp2_pm.c
@@ -970,8 +970,10 @@ static void __dead2 stm32_system_off(void)
 
 #if !STM32MP_M33_TDCID
 	/* Force DDR off */
-	mmio_clrbits_32(rcc_base + RCC_DDRITFCFGR, RCC_DDRITFCFGR_DDRSHR);
 	stm32mp_board_ddr_power_off();
+
+	/* Disable DDRSHR */
+	mmio_clrbits_32(rcc_base + RCC_DDRITFCFGR, RCC_DDRITFCFGR_DDRSHR);
 #endif
 
 	/* Prevent interrupts from spuriously waking up this cpu */
-- 
2.53.0

